### ISSUE TITLE
Accessing contents within cross-domain iframes

### ISSUE THREAD

**Issue by dhowe** [dhowe/AdNauseam#93]
Anyone have any thoughts on this? We would like to be able to access ad content within cross-domain iframes, which is generally disallowed by same-origin-policy. Are there workarounds for within extensions (some posts say yes, but I've yet to find a working example)? I have been able to run content-scripts within all iframes on a page (including cross-domain and about:blank), however not to select elements within these...

 In any case, I'll create simple test pages for the basic and nested cross-domain cases.

https://developer.mozilla.org/en-US/Add-ons/SDK/Guides/Content_Scripts/Cross_Domain_Content_Scripts

**Comment by mushon** [comment]
Are they being blocked by uBlock?

**Comment by dhowe** [comment]
Yes, by default the entire iframe is either blocked or hidden by uBlock. Blocking is disabled entirely for now in adn, as we can can't capture anything blocked, and this testing is being done with iframe hiding disabled, but the general problem is that we cannot, thus far, cannot interrogate any content within any cross-domain iframe

**Comment by dhowe** [comment]
[Here](http://rednoise.org/adntest/iframe-crossdomain.html) is very simple test page.

**Comment by mushon** [comment]
so I say let's go with the simpler way of doing this at this point, which I guess would mean blocking the iFrame and not picking inside it.
We can make this a later enhancement but at this point I don’t think it should be a blocker for the launch of version 2.0.

**Comment by dhowe** [comment]
Here is the error I got when attempting to access the test page(linked above):

![screen shot 2016-03-30 at 6 56 09 pm](https://cloud.githubusercontent.com/assets/737638/14139998/2df6994c-f6a9-11e5-8691-3619ad9c6bbb.png)

**Comment by mushon** [comment]
I'm not sure I understand the technical challenge here. Anyway I was suggesting we take the simpler route (if one exists) and not make this issue a blocker for the release.

**Comment by yansh** [comment]
I don't get any errors from the test page in Chrome.
<img width="967" alt="screen shot 2016-04-08 at 4 28 06 pm" src="https://cloud.githubusercontent.com/assets/1371050/14396677/edc84f14-fda6-11e5-806e-0c47b79008aa.png">

In general, as you mentioned, there are workarounds for known domains. I used background scripts and ```window.postMessage``` but you need to specify the domains in the manifest file. Perhaps using some of "vAPI messaging" can also help.

**Comment by dhowe** [comment]
Errors are not the problem. The problem is, specifically, that we cannot access data inside a cross-domain (often nested) iframe. Thus we cannot capture any ads served this way.

An example: say we have a page (www.thesite.com/homepage.html) which has an cross-domain iframe inside (with src attribute of "ads.google.com?blah", for example), and inside that iframe is a script that creates and loads a 3rd iframe with the ad content (its src attribute is generally "about:blank"). There is no way clean way to access that ad from a content-script.

I have developed a very clunky workaround for the general case as follows: we block the original iframe, capture its source, then create a separate request to that URL via an XMLHttpRequest. When we receive the HTML string response, we convert that string to an off-page DOM tree, then parse that DOM tree (which is no longer subject to cross-domain issues) according to the usual rules/scripts to capture its ads. I have this working, at least for simple test cases.

But clearly it is not ideal in terms of complexity and efficiency, plus it could potentially be blocked if the server distinguishes between normal browser headers and those of the XMLHttpRequest. 

Thus I would prefer another solution -- what workarounds are you referring to?

**Comment by yansh** [comment]
> Thus I would prefer another solution -- what workarounds are you referring to?

Sorry,  I misunderstood the problem. Not sure there is one that is much better than what you have right now.

**Comment by dhowe** [comment]
closing for now, will likely need to revisit post-chrome launch

**Comment by dhowe** [comment]
See also: https://stackoverflow.com/questions/42968473/firefox-webextension-how-to-obtain-and-modify-content-of-cross-domain-iframe
